### ISSUE TITLE
Ordering - Use Cases

### ISSUE THREAD

**Issue by maydanw** [vaticle/typedb#6390]
## Problem to Solve

Ordering in TypeDB is currently far from optimized. There had been a lot of discussions about how it can be solved but this issue comes to address a more primal issue of what are the concrete use cases for ordering.
A suggestion on how to use this issue is to write the use cases in the comments and add a thump up if a use case is similar to yours. 
If there are more similar Use Cases issues there should be a label for them. While opening a related issue it can be linked to the use cases and maybe having a "semi-dialog" in it about what are the use cases it about to solve and what is out of scope.

**Comment by maydanw** [comment]
My use cases are the following:
1. I have in my roadmap a document entity which is built from many entities such as paragraphs, bullets, ... (it is important to me to break it down into independent entities for the next steps and for relations) . Obviously, I need to be able to keep these entities sorted and in a constant order to keep the document to be comprehensible.
2. I have ToDo items entities that the user or the system creates and the user can prioritize them and putting them upper or lower in his view. I need this custom order to be kept.
3. I want to be able to fetch objects "in pages" (e.g. 20-40) from a set of objects with an ordering key (e.g., last modification epoch time)

**Comment by maydanw** [comment]
A document was written by @modeller
[TypeQL_-_List_Behaviour.docx](https://github.com/vaticle/typedb/files/6733188/TypeQL_-_List_Behaviour.docx)

**Comment by causalmodeller** [comment]
These requirements appear to be list-based, assuming an ordered list that can contain any thing (entity, relation, attrubute). thus the requirements above, can be combined with the vector requirements from #6327

**Comment by lolski** [comment]
The use cases you've [described above](https://github.com/vaticle/typedb/issues/6390#issuecomment-870526967) can already be achieved with the `sort`, `offset`, and `limit` functionality [described here](https://docs.vaticle.com/docs/query/get-query). However I agree that they're not fast and should be optimised.

**Comment by lveillard** [comment]
Hello sharing with you an extremely simple example:

Let's say I have 1M ordered nodes (for example ordered with an index in the relation).

If I want to add a new one as the first one... I have to modify a million relations. Just to add one

**Comment by flyingsilverfin** [comment]
Note:
A list that is updatable and indexable is a problem in most systems: consider a `LinkedList` (O(1) update, but O(n) indexed read) vs an ArrayList (O(n) update, but O(1) indexed read).

You can implement a LinkedList in TypeDB with linked relations. An ArrayList can be emulated with an indexed set.

You can do somethink funky where you use a double to index the elements, so that insertion can just choose a double halfway between two existing values, giving you more complex lookups (probably a binary search?) but O(1) inserts. This will work until the doubles run out of precision...

**Comment by lveillard** [comment]
Found an implementation of what you explained. Will implement that in the client side but I think it could be a nice way to have builted-in vectors in typeDB, This is something that would help to differentiate against other graphs DBs that don't have linkedlists. And it does not feel like a rare use case.

https://github.com/brendon/ranked-model/blob/ac4079ebeda2a10b79fe6a488ddff45b0c5a5c46/lib/ranked-model/ranker.rb

**Comment by flyingsilverfin** [comment]
Heads up: sorting on non-string values (and also not using reasoner) is now optimised natively in TypeDB :)

**Comment by maydanw** [comment]
Just to make sure I understand:
So the double option is the best for now?
And with skip list (as predefined ranges) or the like as optimization.
With the coming arithmetic I will be able to push them around and re-spread
if there is a need

**Comment by flyingsilverfin** [comment]
yeah anything to do with ordering will be best optimised with sorted numbers

**Comment by lveillard** [comment]
What about edge ordering (vectors)?
Is there an ETA? Would it be included in the rust refacto? 🤔

**Comment by lveillard** [comment]
Hello @haikalpribadi, I'm talking about ordering on-write, not on query, so probably not the right thread :S

Like storing arrays of values:
$b has codes [1,2,8,4] for instance, in that particular order

or storing vectors of edges:
$a (parent: $p, children: [$1, $2, $3, $4]);
